onnxruntime/test/providers/kernel_compute_test_utils.h

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef ENABLE_TRAINING
+
+#pragma once
+
+#include "test/providers/provider_test_utils.h"
+
+namespace onnxruntime {
+namespace test {
+
+// OpTester creates a graph and run it using InferenceSession, the inputs and outputs are graph inputs and outputs,
+// which cannot be strided tensors for now. This KernelComputeTester creates the input and output OrtValues
```

Comment:
Have you ever considered extending the OpTester for strided tensors inputs, though I agree OpTesters include many things, requires a refactoring, while the functionality is pretty good in my understanding. 

Previously as we talked offline, I have a quick trial to run strited tensor uts, here is a naiive change FYI. https://github.com/microsoft/onnxruntime/commit/64e1ea5459ca4973430451a35d5763ea30124a92#diff-bb3dc26f2f0ac2829a7bc0f37e5af9858bcd0ffb06ee33024027e3be2499223f

Reply:
I have totally opposite opinion on this. OpTester is good as a utility tool, but I think it's very bad for UT. The best UT code should just invoke the tested code without any other impact. But OpTester includes the whole allocation planner and inference session thing. Somethings you just can reach some kernel code path that are intermediate node specific because OpTester's graph have only one node. An extreme case is you want to test a CUDA kernel but end up the inference session has some issue that at the end it runs the CPU code and get pass, but the CUDA code might actually have problem.

Currently our allocation planner cannot support strided tensor as graph output, for any graph output, the allocation type is kOutputAllocated. Actually when they designed this, they didn't expect any strided tensor as graph input, it just works accidentally and might have some unknown behavior. Your code in the PR is this case, you can put strided tensor as graph input, but you cannot test the strided tensor output case. KernelComputeTester actually still have quite some unrelated things, it also requires the graph and has the execution frame, but the execution frame is much simpler than the normal one, and this is the easiest way I could found to do such test.

Reply:
Thanks for the explanation. Few thoughts around what you mention:
1. Specifically for the graph input/output, ideally (and very likely not urgent), we need support as well, right? For cases passing in in-contiguous tensor inputs.
2. Agree having a kernel compute (which is a more fine-grained UT than operator) is a good plus for us. I am curious how we would test strided tensor changes in allocation planner? Correct me if I missed some code.

Reply:
Allocation planner and kernel implementation are independent things. For allocation planner, we have UTs in allocation_planner_test.cc which also cover the strided tensor cases. It will guarantee the tensor is allocated or reused as expected. The kernel UT should focus on the kernel code only, i.e., for strided tensor case, if the output is reused, generate correct strides, otherwise generate contiguous output, it has nothing to do with the allocation planner. Combining them together is an ideal UT, maybe we can call it integration test, and if we want such test (if you think it's necessary), we can add it. In fact, for ORTModule, the tests in those py files are actually similar to such case and we can add (or maybe we already have, for example the BERT test) cases to cover such cases.

It' easy to support strided tensor as input. I actually don't figure out a good way to support strided tensor as graph output in current memory owner logic of our current tensor, it's not possible to pass out an intermediate tensor out as graph output. But even in the future we can support it, the UTs for them are still in allocation_planner_test.cc, and have nothing to do with the kernel implementation UTs.